### PR TITLE
Develop biometric

### PR DESCRIPTION
**Pull request by Ogerets**

### CODE REVIEW

File: Source/EThree.swift [hunk 1]
```diff
@@ -73,33 +73,89 @@ import VirgilCrypto
...
     /// - Important: identity should be the same as in JWT generated at server side
+    /// - Note: To use TouchId or FaceId, in addion `biometricProtection` to being enabled,
+    ///  `NSFaceIDUsageDescription` key should be included in your app’s Info.plist file.
+    /// See [here](https://developer.apple.com/documentation/localauthentication/logging_a_user_into_your_app_with_face_id_or_touch_id)
+    /// more info
+    /// - Note: `biometricalProtection` flag does not resave all entries with new options.
+    /// Use `setBiometricalProtection` method for this purpose
     @objc public convenience init(identity: String,
                                   tokenCallback: @escaping RenewJwtCallback,
+                                  biometricProtection: Bool,
```
Comment (SanjoDeundiak) [inline Source/EThree.swift:121]: Too many parameters
1. Add some params struct
2. Add possibility to configure E3Kit using config file

File: Source/PublicAPI/EThree+Authorization.swift [hunk 2]
```diff
@@ -73,14 +75,24 @@ extension EThree {
         }
     }
 
+#if os(iOS)
+    /// Updates keychain entries with new options
+    ///
+    /// - Parameter set: private key will be accessed only using biometrics or passcode if true
+    /// - Throws: corresponding error
+    @objc public func setBiometricProtection(to set: Bool) throws {
```
Comment (SanjoDeundiak) [inline Source/PublicAPI/EThree+Authorization.swift]: Bad argument naming

Comment (SanjoDeundiak) [reply Source/PublicAPI/EThree+Authorization.swift]: What's the use case for this method?

File: Source/Storages/Local/LocalKeyStorage.swift [hunk 3]
```diff
@@ -39,33 +39,129 @@ import VirgilSDK
 
 internal class LocalKeyStorage {
     internal let identity: String
-    private let crypto: VirgilCrypto
+    internal let crypto: VirgilCrypto
+
+    private var keyPair: VirgilKeyPair?
```
Comment (SanjoDeundiak) [inline Source/Storages/Local/LocalKeyStorage.swift]: Should add possibility to reauthenticate user using biometrics after timer and/or after minimizing the app

File: Source/Storages/Local/LocalKeyStorage.swift [hunk 4]
```diff
@@ -39,33 +39,129 @@ import VirgilSDK
...
+    internal required init(identity: String,
+                           crypto: VirgilCrypto,
+                           keychainStorage: KeychainStorage,
+                           options: KeychainQueryOptions = KeychainQueryOptions()) throws {
         self.identity = identity
         self.crypto = crypto
         self.keychainStorage = keychainStorage
+        self.options = options
+
+        try self.loadKeyPair()
```
Comment (SanjoDeundiak) [inline Source/Storages/Local/LocalKeyStorage.swift]: Load key-pair only when you need it

File: Source/Storages/Local/LocalKeyStorage.swift [hunk 5]
```diff
@@ -39,33 +39,129 @@ import VirgilSDK
...
+        _ = try self.keychainStorage.store(data: data,
+                                           withName: self.backupName,
+                                           meta: nil)
+    }
+
+    private func deleteBackup() throws {
+        try self.keychainStorage.deleteEntry(withName: self.backupName)
+    }
+
+    private func applyBackup() throws {
```
Comment (SanjoDeundiak) [inline Source/Storages/Local/LocalKeyStorage.swift]: Bad function naming

File: Source/Storages/Local/LocalKeyStorage.swift (hunk 3 above)
Comment (SanjoDeundiak) [reply Source/Storages/Local/LocalKeyStorage.swift]: Need to make sure that key is not stored anywhere outside of this class for duration more than needed for one cryptographic operation

**Review by SanjoDeundiak — CHANGES_REQUESTED** [review]

File: Source/PublicAPI/EThree+Authorization.swift (hunk 2 above)
Comment (Ogerets) [reply Source/PublicAPI/EThree+Authorization.swift]: If user wants to stop using biometric protection for key, he needs some method to resave key with new options. Apps with biometric protection provide such functionality

File: Source/Storages/Local/LocalKeyStorage.swift (hunk 4 above)
Comment (Ogerets) [reply Source/Storages/Local/LocalKeyStorage.swift]: User opens app, enters chat, tries to send new message and suddenly face id check appears? Doesn't seem to be good UX

File: Source/EThree.swift (hunk 1 above)
Comment (Ogerets) [reply Source/EThree.swift:121]: 1. How would this help?

File: Source/Storages/Local/LocalKeyStorage.swift (hunk 3 above)
Comment (Ogerets) [reply Source/Storages/Local/LocalKeyStorage.swift]: >Should add possibility to reauthenticate user using biometrics after timer and/or after minimizing the app

Timer is bad for UX. Suddenly appeared check is weird. Will investigate possibility to observe minimizing app.

**Pull request closed without merging**